src/virtual_allocator.rs

```diff
@@ -162,6 +187,8 @@ unsafe impl Allocator for VirtualMemoryManager {
         if size > self.block_size_in_bytes {
             let num_blocks = size / self.block_size_in_bytes;
             if let Some(range) = self.find_adjacent_free_blocks(num_blocks) {
+                self.maximum_tail_index
+                    .fetch_max(range.end, Ordering::SeqCst);
```

Comment:
i think this is fine and afaict `range.end` is consistent with `index + 1` below.
My only minor suggestion is we could make it an optional feature. If we don't care about stats tracking, we don't need this in the hot path for every allocate call.
If we consider thread safety out of scope for the moment, we also don't need the value to be atomic.

Reply:
Feature implemented by https://github.com/matter-labs/era-shivini/pull/7/commits/674a7f4bcaed13595347c100f981a4500373f1a7.
I want to keep the atomic to keep in line with the rest of the Allocator implementation, on it's own this one field and it's usage is thread safe.